### ISSUE TITLE
Support textDocument/codeLens

### ISSUE THREAD

**Issue by ayoub-benali** [sublimelsp/LSP#781]
See spec: https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#textDocument_codeLens

RLS and Metals rely on this for running tests at least

I am still wondering if this possible given the current sublime API or we should give feedback regarding the new versions.

**Comment by tomv564** [comment]
Good UI challenge - looks like VS Code uses a UI that we could build using phantoms
<img width="392" alt="Screenshot 2019-11-18 at 21 13 45" src="https://user-images.githubusercontent.com/4395832/69089297-63697480-0a48-11ea-8f66-8de2c4c7dc96.png">

**Comment by predragnikolic** [comment]
Like Tom said, that is doable said with phantoms.

![ddd](https://user-images.githubusercontent.com/22029477/69093621-18069480-0a4f-11ea-929e-1fd43923db62.png)

here is how the code for that (https://github.com/predragnikolic/SIDE/blob/2b38006e015b229ab603f016d83e2a1f6060edcf/features/code_lens.py#L26-L80)

**Comment by ayoub-benali** [comment]
thanks for the example.
Now that I know how we could display the code lens, I wonder how often the client should send a `textDocument/codeLens` request. 
Would it be okay to do it after each `textDocument/didChange` ?
Also we need probably a way to save the state of each file with it's array of CodeLens

**Comment by predragnikolic** [comment]
You can have a look at the code-lens [WIP](https://github.com/sublimelsp/LSP-elm/tree/code-lens)  I did for the elm language server. 

Long story short, 
I copied the implementation for the color request class, 
also modified the LSP plugin to support code lens in client capabilities, 
defined a code_lens method on the Request class... I got a response from the server, but didn't knew what to do with it. :)

I think the code lens should fire on these two events:
- on_activated_async,
- on_modified_async (I don't know how expensive the request will be if we send it on each keystroke)

**Comment by ayoub-benali** [comment]
@predragnikolic  thanks for the tip! :)
> I got a response from the server, but didn't knew what to do with it.

I suppose we are supposed to display a ST phantom with the code lens command attached to it

**Comment by rwols** [comment]
Wouldn't it be better as a sort of hover-popup thing? I thought it was supposed to look like Visual Studio's implementation:

https://docs.microsoft.com/en-us/visualstudio/ide/find-code-changes-and-other-history-with-codelens?view=vs-2019

**Comment by rwols** [comment]
I guess there are two parts to this. The first is the tiny text above/under a symbol, the second is the popup. We'd have to experience how intrusive phantoms would be, as they move the text below/under them.  I think @predragnikolic has the right approach.

**Comment by ayoub-benali** [comment]
What would be the popup for ? showing the button issuing a `codeLens/resolve` ?

**Comment by rwols** [comment]
@ayoub-benali for showing this kind of rich information:

https://docs.microsoft.com/en-us/visualstudio/ide/media/codelens-overview.png?view=vs-2019

**Comment by predragnikolic** [comment]
Here is how it looks in VS code.
![good](https://user-images.githubusercontent.com/22029477/69831874-b18e2d00-122b-11ea-9c0b-6858ed14a18b.gif)

I guess it depends on what the best approach is for handling cases when a certain code lens is clicked. 

For example,
- If a user clicks on `4 references` code lens.
I would open up the bottom panel and list all references found, like do for the `textDocument/references` request. And list all reference there.

![2019-11-28-221156_635x459_scrot](https://user-images.githubusercontent.com/22029477/69832214-64ab5600-122d-11ea-81dd-b0dcb1bb86d1.png)

- If we had a `run test` code lens.
I would also open up the bottom panel and list the response there for the successful/failed tests.

Maybe there will be a code lens that should be handled differently. Like for example showing it in a hover popup(although I am not sure if we can listen to on_hover on a phantom?).

**Comment by predragnikolic** [comment]
> I suppose we are supposed to display a ST phantom with the code lens command attached to it.

Here is the spec for the code lens. Note that the command is optional.
```
interface CodeLens {
	/**
	 * The range in which this code lens is valid. Should only span a single line.
	 */
	range: Range;

	/**
	 * The command this code lens represents.
	 */
	command?: Command;

	/**
	 * A data entry field that is preserved on a code lens item between
	 * a code lens and a code lens resolve request.
	 */
	data?: any
}
```

<details>
  <summary>here is the response I got from the elm language server :)</summary>

```
[
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": True,
            "codeLensType": "exposed",
            "nameNode": "main"
        },
        "range": {
            "end": { "character": 63, "line": 12 },
            "start": { "character": 0, "line": 11 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": False,
            "codeLensType": "exposed",
            "nameNode": "Model"
        },
        "range": {
            "end": { "character": 3, "line": 23 },
            "start": { "character": 0, "line": 19 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": True,
            "codeLensType": "exposed",
            "nameNode": "init"
        },
        "range": {
            "end": { "character": 12, "line": 26 },
            "start": { "character": 0, "line": 26 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": False,
            "codeLensType": "exposed",
            "nameNode": "Msg"
        },
        "range": {
            "end": { "character": 24, "line": 38 },
            "start": { "character": 0, "line": 35 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": True,
            "codeLensType": "exposed",
            "nameNode": "update"
        },
        "range": {
            "end": { "character": 30, "line": 41 },
            "start": { "character": 0, "line": 41 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": True,
            "codeLensType": "exposed",
            "nameNode": "view"
        },
        "range": {
            "end": { "character": 24, "line": 58 },
            "start": { "character": 0, "line": 58 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": True,
            "codeLensType": "exposed",
            "nameNode": "viewInput"
        },
        "range": {
            "end": { "character": 69, "line": 68 },
            "start": { "character": 0, "line": 68 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "isFunction": True,
            "codeLensType": "exposed",
            "nameNode": "viewValidation"
        },
        "range": {
            "end": { "character": 34, "line": 73 },
            "start": { "character": 0, "line": 73 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 16, "line": 19 },
            "start": { "character": 11, "line": 19 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 8, "line": 35 },
            "start": { "character": 5, "line": 35 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 63, "line": 12 },
            "start": { "character": 0, "line": 11 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 12, "line": 26 },
            "start": { "character": 0, "line": 26 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 30, "line": 41 },
            "start": { "character": 0, "line": 41 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 24, "line": 58 },
            "start": { "character": 0, "line": 58 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 69, "line": 68 },
            "start": { "character": 0, "line": 68 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 34, "line": 73 },
            "start": { "character": 0, "line": 73 }
        }
    },
    {
        "data": {
            "uri": "file:///home/predrag/Documents/elm/elm-architecture-tutorial/examples/03-form.elm",
            "codeLensType": "referenceCounter"
        },
        "range": {
            "end": { "character": 11, "line": 0 },
            "start": { "character": 7, "line": 0 }
        }
    }
]
```
</details>
<details>
  <summary>Here is the content of the file</summary>
module Main exposing (..)
import Browser
import Html exposing (..)
import Html.Attributes exposing (..)
import Html.Events exposing (onInput)

 

-- MAIN


main =
  Browser.sandbox { init = init, update = update, view = view }



-- MODEL


type alias Model =
  { name : String
  , password : String
  , passwordAgain : String
  }


init : Model
init =
  Model "" "" ""



-- UPDATE


type Msg
  = Name String
  | Password String
  | PasswordAgain String


update : Msg -> Model -> Model
update msg model =
  case msg of
    Name name ->
      { model | name = name }

    Password password ->
      { model | password = password }

    PasswordAgain password ->
      { model | passwordAgain = password }



-- VIEW


view : Model -> Html Msg
view model =
  div []
    [ viewInput "text" "Name" model.name Name
    , viewInput "password" "Password" model.password Password
    , viewInput "password" "Re-enter Password" model.passwordAgain PasswordAgain
    , viewValidation model
    ]


viewInput : String -> String -> String -> (String -> msg) -> Html msg
viewInput t p v toMsg =
  input [ type_ t, placeholder p, value v, onInput toMsg ] []


viewValidation : Model -> Html msg
viewValidation model =
  if model.password == model.passwordAgain then
    div [ style "color" "green" ] [ text "OK" ]
  else
    div [ style "color" "red" ] [ text "Passwords do not match!" ]
</details>


I should probably reach out for help from the maintainer of the elm language server, so I could better understand the payload. 

There is probably something I am missing about the code lens implementation. I didn't understand the [codeLens/resolve](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#codeLens_resolve) request. I am not sending a `codeLens/resolve`, maybe I should?

**Comment by rwols** [comment]
I think once the user selects an item from the initial code lens request, we must make a codelens/resolve request with the selected codelens item as params?

**Comment by ayoub-benali** [comment]
yeah that is my understating as well, but `codelens/resolve` can be sent only when there is a `command` otherwise it is just contextual data to be displayed

**Comment by predragnikolic** [comment]
![2019-12-18-231818_621x621_scrot](https://user-images.githubusercontent.com/22029477/71128233-1286a080-21ed-11ea-80cd-1a98451e75e7.png)

A quick update. 
I finally figured it out how the code lens request works.

First we need to send a `textDocument/codeLens` request.
`--> textDocument/codeLens`
It returns us an array of code lenses. 
`<-- CodeLens[]`

I read this [doc](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#textDocument_codeLens):

> * A code lens represents a command that should be shown along with
>  * source text, like the number of references, a way to run tests, etc.
>  *
> * A code lens is _unresolved_ when no command is associated to it. For performance
> * reasons the creation of a code lens and resolving should be done in **two stages**.


so after we got the array of code lenses.
for each code lens we need to send a `‘codeLens/resolve’` request,
 passing it a code lens as a param.

And finally we get the code lens response that we want.

Although I got the response back, I am having trouble with getting the user experience right with phantoms. But never the less, maybe someone will find it useful and build on top of it. The code is [here](https://github.com/sublimelsp/LSP-elm/blob/code-lens/codeLens.py).

**Comment by rwols** [comment]
@TheSecEng weren't you playing around with this?

**Comment by TheSecEng** [comment]
I am/was. Haven't touched it in a few weeks, probably revisit this weekend to see how far along I got !

**Comment by rwols** [comment]
Resolved with #1477.

- Rendered as annotations.
- Lazily resolves unresolved lenses visible in the viewport. Unfortunately, there's no "viewport changed" callback in ST3/ST4. So, best we can do is check the viewport on every selection changed event.
- These code lenses are **highly** dependent on client glue code, so **only sessions backed by an AbstractPlugin** will request code lenses. Your plugin is expected to handle the **highly custom commands** by using the `on_pre_server_command` callback.
